### PR TITLE
Add method names next to reloc adrp/add in disasm

### LINKED ISSUE #35271: [Arm64] Fix JIT disasm of reloc adrp/add to be more explicit
In R2R arm64 code, we often see code sequences like:
```
adrp    x11, [RELOC #0x1f40fb94a68]
add     x11, x11, #0
```

This is confusing because it looks like the `add x11, x11, #0` instruction is useless. But it is the second instruction in an `adrp/add` pair that gets a relocation.

Change the disassembly for this `add` instruction to indicate it is getting a relocation.

For example, output:
```
adrp    x11, [HIGH RELOC #0x1f40fb94a68]
add     x11, x11, [LOW RELOC #0x1f40fb94a68]
```

### PR DESCRIPTION
**Pull request by kunalspathak**
Added method names next to the disasm code of adrp/add to reflect the target method. I tried checking every possible code path that generated adrp/add for `System.Private.Corelib.dll` and passed `methodHandle` that can be queried during disassembly printing to get the target method name. For places where `methodHandle` was not available (or I was not sure what to use) added `MethodHandleType enum` to tell disassembler what type of target it is. I made sure that the values of these enum don't overlap with those of helper or user function. So for `System.Private.Corelib`, we should never see `Unknown` as the target method name. However just before sending out this PR, I ran crossgen over framework libraries and I see bunch of places where we are seeing `Unknown`. Finding those code path should be easy because in a separate branch I added logging at around 60 places to track those code path. 

~If I get time, I will try to fix those code path as well, but just wanted to send out PR.~ I added one last `FieldAccess` type after which we have valid target name for all generated crossgened code for framework libraries.

Below are type of target methods that are identified and printed in disassembly.

| Scenario                                       | Count |
|------------------------------------------------|-------|
| Helper call                                    | 10813 |
| Virtual stub address                           | 2730  |
| Indirect cell address                          | 80486 |
| GS cookie check                                | 521   |
| Setting GS cookie                              | 381   |
| Static field access                            | 101   |
| Direct call (indirection access)               | 5937  |
| Nonvirtual PInvoke (double indirection access) | 201   |
|  Nonvirtual PInvoke (indirection access)       | 4     |
| String literal lookup                          | 15984 |
| Static lookup tree                             | 2112  |
| Runtime lookup tree                            | 443   |
| Initialize array elements                      | 87    |

With this change, here is the sample asm:

```asm
        2A1403E2          mov     w2, w20
        B9008FA2          str     w2, [fp,#140]
        9000000B          adrp    x11, [HIGH RELOC #0x29747e95ca8]      // [BCrypt:BCryptGenRandom(long,long,int,int):int]
        9100016B          add     x11, x11, [LOW RELOC #0x29747e95ca8]
        F9003FAB          str     x11, [fp,#120]
        D2800000          mov     x0, #0
        F9003BA0          str     x0, [fp,#112]
        52800043          mov     w3, #2
        B9008BA3          str     w3, [fp,#136]
        910063A0          add     x0, fp, #24   // [V14 PInvokeFrame]
        90000004          adrp    x4, [HIGH RELOC #0x29747e96080]      // [CORINFO_HELP_JIT_PINVOKE_BEGIN]
        91000084          add     x4, x4, [LOW RELOC #0x29747e96080]
        F9400084          ldr     x4, [x4]
```

Fixes: https://github.com/dotnet/runtime/issues/35271

### CODE REVIEW

**Comment by kunalspathak** [comment]
@dotnet/jit-contrib

File: src/coreclr/src/jit/emitarm64.cpp [hunk 1]
```diff
@@ -12082,8 +12091,40 @@ void emitter::emitDispIns(
...
+                    switch (methodHandle)
+                    {
+                        case GenTreeIntCon::MethodHandleType::Unknown:
+                            targetName = "Unknown";
+                            break;
+                        case GenTreeIntCon::MethodHandleType::StringLiteralNode:
+                            targetName = "Access_To_StringLiteralNode";
+                            break;
+                        case GenTreeIntCon::MethodHandleType::StaticLookupTree:
+                            targetName = "Static_LookupTree";
```
Comment (kunalspathak) [inline src/coreclr/src/jit/emitarm64.cpp:12107]: I was not sure what is the right term to describe `StaticLookupTree` and `RuntimeLookupTree`. Any suggestions?

File: src/coreclr/src/jit/emitarm64.cpp [hunk 2]
```diff
@@ -12066,6 +12074,7 @@ void emitter::emitDispIns(
             printf("[");
             if (id->idAddr()->iiaIsJitDataOffset())
             {
+                printf("[");
```
Comment (BruceForstall) [inline src/coreclr/src/jit/emitarm64.cpp:12077]: What is this for? Won't we end up with `[[...]`?

File: src/coreclr/src/jit/gentree.cpp [hunk 3]
```diff
@@ -5928,11 +5928,19 @@ GenTree* Compiler::gtNewStringLiteralNode(InfoAccessType iat, void* pValue)
         case IAT_PVALUE: // The value needs to be accessed via an indirection
             // Create an indirection
             tree = gtNewIndOfIconHandleNode(TYP_REF, (size_t)pValue, GTF_ICON_STR_HDL, false);
+#ifdef DEBUG
+            tree->gtGetOp1()->AsIntCon()->gtMethodHandle =
+                (CORINFO_METHOD_HANDLE)GenTreeIntCon::MethodHandleType::StringLiteralNode;
```
Comment (BruceForstall) [inline src/coreclr/src/jit/gentree.cpp:5933]: This is a case where hopefully (in a follow-up) we could store the actual string handle and then display the actual string (or at least a prefix, for long strings) in the asm output.

Don't you need another case for the `IAT_VALUE` case above?

File: src/coreclr/src/jit/gentree.h [hunk 4]
```diff
@@ -2959,6 +2959,23 @@ struct GenTreeIntCon : public GenTreeIntConCommon
     // sequence of fields.
     FieldSeqNode* gtFieldSeq;
 
+#ifdef DEBUG
+    // If the value represents target address, holds the method handle to that target which is used
+    // to fetch target method name and display in the disassembled code.
+    CORINFO_METHOD_HANDLE gtMethodHandle = (CORINFO_METHOD_HANDLE)Unknown;
+
+    // Enum values are such that they don't overlap with helper's or user function's method handle.
+    enum MethodHandleType{Unknown                  = 2,
```
Comment (BruceForstall) [inline src/coreclr/src/jit/gentree.h:2968]: The `gtFlags` already contains `GTF_ICON_*_HDL` (everything under `GTF_ICON_HDL_MASK`). So is that sufficient, without needing this enum? Then, this `gtMethodHandle` could potentially be a union of different handle types that is interpreted based on that `GTF_*` value?

File: src/coreclr/src/jit/morph.cpp [hunk 5]
```diff
@@ -6094,6 +6097,10 @@ GenTree* Compiler::fgMorphField(GenTree* tree, MorphAddrContext* mac)
             {
                 offsetNode = gtNewIndOfIconHandleNode(TYP_I_IMPL, (size_t)tree->AsField()->gtFieldLookup.addr,
                                                       GTF_ICON_FIELD_HDL, false);
+#ifdef DEBUG
+                offsetNode->gtGetOp1()->AsIntCon()->gtMethodHandle =
+                    (CORINFO_METHOD_HANDLE)GenTreeIntCon::MethodHandleType::FieldAccess;
```
Comment (BruceForstall) [inline src/coreclr/src/jit/morph.cpp:6102]: Is this a place where we should save `gtFldObj`?

File: src/coreclr/src/jit/emitarm64.cpp (hunk 2 above)
Comment (kunalspathak) [reply src/coreclr/src/jit/emitarm64.cpp:12077]: Yeah, that is the left over code of something that I was trying. Will remove it.

File: src/coreclr/src/jit/morph.cpp (hunk 5 above)
Comment (kunalspathak) [reply src/coreclr/src/jit/morph.cpp:6102]: Correct.

File: src/coreclr/src/jit/gentree.h (hunk 4 above)
Comment (kunalspathak) [reply src/coreclr/src/jit/gentree.h]: Yes, didn't realize that I can just use `gtFlags`.

File: src/coreclr/src/jit/gentree.cpp (hunk 3 above)
Comment (kunalspathak) [reply src/coreclr/src/jit/gentree.cpp]: Yes, fixed it although, today `eeGetCPString()` method used to get string literal from `stringHandle` is not supported on linux/arm. I have added a note.

Comment (kunalspathak) [reply src/coreclr/src/jit/gentree.cpp]: Also, I didn't notice the location of `IAT_VALUE` being used under `adrp` so didn't add. But I think it makes sense to add it there.

File: src/coreclr/src/jit/emitarm64.cpp (hunk 1 above)
Comment (kunalspathak) [reply src/coreclr/src/jit/emitarm64.cpp]: Not needed anymore.

**Review by BruceForstall — APPROVED** [review]
LGTM

**Pull request merged**